src/launcherext.ts

```diff
@@ -0,0 +1,169 @@
+const Me = imports.misc.extensionUtils.getCurrentExtension();
+
+const { evaluate } = Me.imports.math.math;
+const { spawnCommandLine } = imports.misc.util;
+
+const { GLib, St } = imports.gi;
+
+import * as log from 'log';
+import * as once_cell from 'once_cell';
+
+import type { Ext } from 'extension';
+import type { Search, SearchResult } from 'search';
+
+const LAUNCHER_ICON_SIZE = 34;
+
+const TERMINAL = new once_cell.OnceCell<string>();
+
+export type LauncherExtension = {
+    // Mode Prefix for launcher
+    prefix: string;
+
+    // Name for debugging
+    name: string;
+
+    /**
+     * Initializes the launcher extension
+     * @param ext Extension settings
+     * @param search Launcher instance
+     */
+    init(ext: Ext, search: Search): LauncherExtension;
+
+    /**
+     * Perform apply action
+     * @param text The currently typed text in input
+     * @param index The currently selected index
+     * @returns true to keep showing the launcher after apply, false to dismiss launcher
+     */
+    apply(text: string, index: number): boolean;
+
+    /**
+     * Gets search results from the launcher extension if search is supported
+     * @param text The currently typed text in input
+     * @returns An array of tuples containing string to display
+     */
+    search_results?: (text: string) => Array<SearchResult> | null;
+}
+
+export class CalcLauncher implements LauncherExtension {
+    prefix = '=';
+    name = 'calc';
+    ext?: Ext;
+    search?: Search;
+
+    init(ext: Ext, search: Search): this {
+        this.ext = ext;
+        this.search = search;
+
+        return this;
+    }
+
+    apply = (expr: string): boolean => {
+        const value: string = evaluate(expr).toString();
+        log.info(`${expr} = ${value}`);
+        if (!this.search) {
+            log.error("init was never called");
+        }
+
+        this.search?.set_text(`=${value}`);
+
+        return true;
+    }
+
+    search_results = (expr: string): Array<SearchResult> | null => {
+        const item: SearchResult =
+        {
+            display: `=${evaluate(expr).toString()}`,
```

Comment:
Doing this as the user types actually generates a lot of logs...

Reply:
![image](https://user-images.githubusercontent.com/661373/81134924-45bf9700-8f24-11ea-81de-57a31cbe37f1.png)